File: laondry-system-management-core-api/app/Models/OrderService/Order.php
<?php

namespace App\Models\OrderService;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Concerns\HasUuids;
use App\Models\Auth\User;

class Order extends Model
{
    use HasFactory, HasUuids, SoftDeletes;

    protected $fillable = [
        'order_no',
        'customer_id',
        'outlet_id',
        'status',
        'order_type',
        'requested_pickup_at',
        'promised_at',
        'pickup_address',
        'delivery_address',
        'total_weight',
        'total_piece',
        'subtotal',
        'discount',
        'tax',
        'delivery_fee',
        'grand_total',
        'external_invoice_id',
        'external_payment_id',
        'notes',
        'created_by',
        'updated_by'
    ];
    protected $casts = [
        'requested_pickup_at' => 'datetime',
        'promised_at' => 'datetime',
        'total_weight' => 'decimal:2',
        'subtotal' => 'decimal:2',
        'discount' => 'decimal:2',
        'tax' => 'decimal:2',
        'delivery_fee' => 'decimal:2',
        'grand_total' => 'decimal:2',
    ];

    public function customer()
    {
        return $this->belongsTo(User::class, 'customer_id');
    }
    public function outlet()
    {
        return $this->belongsTo(Outlet::class);
    }
    public function statusRef()
    {
        return $this->belongsTo(OrderStatus::class, 'status', 'code');
    }

    public function items()
    {
        return $this->hasMany(OrderItem::class);
    }
    public function logs()
    {
        return $this->hasMany(OrderStatusLog::class);
    }

    public function creator()
    {
        return $this->belongsTo(User::class, 'created_by');
    }
    public function updater()
    {
        return $this->belongsTo(User::class, 'updated_by');
    }

    // helper hitung total
    public function recalcTotals(): void
    {
        $subtotal = $this->items()->sum('line_total');
        $this->subtotal = $subtotal;
        $this->grand_total = $subtotal - $this->discount + $this->tax + $this->delivery_fee;
    }
}

File: laondry-order-service/internal/entity/order.go
package entity

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type Order struct {
	ID                uuid.UUID      `gorm:"type:uuid;primary_key" json:"id"`
	CustomerID        uuid.UUID      `gorm:"type:uuid;not null;index" json:"customer_id"`
	OutletID          uuid.UUID      `gorm:"type:uuid;not null;index" json:"outlet_id"`
	Status            string         `gorm:"type:varchar(30);not null;default:'NEW'" json:"status_code"` // Mobile expects status_code
	OrderNo           string         `gorm:"type:varchar(30);not null;unique" json:"order_no"`
	OrderType         string         `gorm:"type:varchar(20);not null;default:'DROPOFF'" json:"order_type"`
	RequestedPickupAt *time.Time     `json:"requested_pickup_at"`
	PromisedAt        *time.Time     `json:"promised_at"`
	PickupAddress     *string        `gorm:"type:varchar(255)" json:"pickup_address"`
	DeliveryAddress   *string        `gorm:"type:varchar(255)" json:"delivery_address"`
	TotalWeight       float64        `gorm:"type:decimal(8,2);default:0" json:"total_weight"`
	TotalPiece        int            `gorm:"default:0" json:"total_piece"`
	Subtotal          float64        `gorm:"type:decimal(12,2);default:0" json:"subtotal"`
	Discount          float64        `gorm:"type:decimal(12,2);default:0" json:"discount_amount"` // Mobile expects discount_amount
	Tax               float64        `gorm:"type:decimal(12,2);default:0" json:"tax_amount"`      // Mobile expects tax_amount
	DeliveryFee       float64        `gorm:"type:decimal(12,2);default:0" json:"delivery_fee"`
	GrandTotal        float64        `gorm:"type:decimal(12,2);default:0" json:"total"` // Mobile expects total
	ExternalInvoiceID *string        `gorm:"type:varchar(100)" json:"external_invoice_id"`
	ExternalPaymentID *string        `gorm:"type:varchar(100)" json:"external_payment_id"`
	Notes             *string        `gorm:"type:text" json:"notes"`
	CreatedBy         *uuid.UUID     `gorm:"type:uuid" json:"created_by"`
	UpdatedBy         *uuid.UUID     `gorm:"type:uuid" json:"updated_by"`
	CreatedAt         time.Time      `gorm:"not null" json:"created_at"`
	UpdatedAt         time.Time      `gorm:"not null" json:"updated_at"`
	DeletedAt         gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`
	IsExpress         bool           `gorm:"-" json:"is_express"` // Virtual field, computed from items

	Customer   *User            `gorm:"foreignKey:CustomerID;references:ID" json:"customer,omitempty"`
	Outlet     *Outlet          `gorm:"foreignKey:OutletID;references:ID" json:"outlet,omitempty"`
	Items      []OrderItem      `gorm:"foreignKey:OrderID;constraint:OnDelete:CASCADE" json:"items,omitempty"`
	StatusLogs []OrderStatusLog `gorm:"foreignKey:OrderID;constraint:OnDelete:CASCADE" json:"status_logs,omitempty"`
}

func (Order) TableName() string {
	return "orders"
}

func (o *Order) BeforeCreate(tx *gorm.DB) error {
	if o.ID == uuid.Nil {
		o.ID = uuid.New()
	}
	return nil
}

// MarshalJSON adds computed fields for mobile compatibility
func (o *Order) MarshalJSON() ([]byte, error) {
	type Alias Order

	// Add outlet_name from preloaded Outlet
	outletName := ""
	if o.Outlet != nil {
		outletName = o.Outlet.Name
	}

	// Add status_name (map status code to display name)
	statusName := getStatusName(o.Status)

	// Check if any item is express
	isExpress := false
	for range o.Items {
		// For now, we assume all items have same express status
		// TODO: Store is_express in order_items table
		break
	}
	o.IsExpress = isExpress

	return json.Marshal(&struct {
		*Alias
		OutletName *string `json:"outlet_name"`
		StatusName string  `json:"status_name"`
	}{
		Alias:      (*Alias)(o),
		OutletName: &outletName,
		StatusName: statusName,
	})
}

func getStatusName(status string) string {
	statusMap := map[string]string{
		"NEW":         "Baru",
		"PENDING":     "Menunggu",
		"IN_PROGRESS": "Diproses",
		"COMPLETED":   "Selesai",
		"CANCELED":    "Dibatalkan",
		"CANCELLED":   "Dibatalkan",
	}
	if name, ok := statusMap[status]; ok {
		return name
	}
	return status
}
